File: frontend/src/types/prediction.ts
export interface HorizonPrediction {
  direction: 'up' | 'down';
  probability: number;
}

/** ML prediction result */
export interface PredictionResult {
  status: string;
  gold_type: string;
  model_trained: string | null;
  message: string | null;
  current_date: string | null;
  current_close: number | null;
  direction: 'up' | 'down' | null;
  direction_probability: number | null;
  predicted_return_7d_pct: number | null;
  predicted_price_7d: number | null;
  confidence_low: number | null;
  confidence_high: number | null;
  horizon_predictions: Record<string, HorizonPrediction>;
  consensus: string | null;
  consensus_label: string | null;
  votes_up: number | null;
  total_horizons: number | null;
  feature_contributions: { feature: string; importance: number }[];
}

/** Model training result */
export interface TrainingResult {
  status: string;
  gold_type: string | null;
  message: string | null;
  model_path: string | null;
  direction_accuracy: number | null;
  direction_precision: number | null;
  direction_recall: number | null;
  direction_f1: number | null;
  price_mae_pct: number | null;
  price_rmse_pct: number | null;
  consensus_strong_pct: number | null;
  consensus_accuracy: number | null;
  feature_importance: { feature: string; importance: number }[];
  train_samples: number | null;
  test_samples: number | null;
  train_date_range: string | null;
}

/** Model metadata */
export interface ModelInfo {
  gold_type: string | null;
  training_date: string | null;
  train_samples: number | null;
  test_samples: number | null;
  dir_accuracy: number | null;
  rmse_pct: number | null;
  consensus_pct: number | null;
  train_date_range: string | null;
  feature_count: number | null;
  top_features: { feature: string; importance: number }[];
  file_size_mb: number | null;
}

File: backend/app/schemas/prediction.py
"""Pydantic schemas for prediction API responses."""

from pydantic import BaseModel


class FeatureContribution(BaseModel):
    feature: str
    importance: float


class HorizonPrediction(BaseModel):
    direction: str  # "up" or "down"
    probability: float


class PredictionResponse(BaseModel):
    status: str
    gold_type: str
    model_trained: str | None = None
    message: str | None = None
    current_date: str | None = None
    current_close: float | None = None
    # 7d summary (backward compat)
    direction: str | None = None
    direction_probability: float | None = None
    predicted_return_7d_pct: float | None = None
    predicted_price_7d: float | None = None
    confidence_low: float | None = None
    confidence_high: float | None = None
    # Multi-horizon
    horizon_predictions: dict[str, HorizonPrediction] = {}
    consensus: str | None = None  # "strong_up" | "strong_down" | "lean_up" | "lean_down" | "mixed"
    consensus_label: str | None = None  # "一致看涨" | "一致看跌" | "偏多" | "偏空" | "分歧"
    votes_up: int | None = None
    total_horizons: int | None = None
    feature_contributions: list[FeatureContribution] = []


class FeatureImportanceItem(BaseModel):
    feature: str
    importance: float


class TrainingResultResponse(BaseModel):
    status: str
    gold_type: str | None = None
    message: str | None = None
    model_path: str | None = None
    direction_accuracy: float | None = None
    direction_precision: float | None = None
    direction_recall: float | None = None
    direction_f1: float | None = None
    price_mae_pct: float | None = None
    price_rmse_pct: float | None = None
    consensus_strong_pct: float | None = None
    consensus_accuracy: float | None = None
    feature_importance: list[FeatureImportanceItem] = []
    train_samples: int | None = None
    val_samples: int | None = None
    test_samples: int | None = None
    train_date_range: str | None = None


class ModelInfoResponse(BaseModel):
    gold_type: str | None = None
    training_date: str | None = None
    train_samples: int | None = None
    test_samples: int | None = None
    dir_accuracy: float | None = None
    rmse_pct: float | None = None
    consensus_pct: float | None = None
    train_date_range: str | None = None
    feature_count: int | None = None
    top_features: list[FeatureImportanceItem] = []
    file_size_mb: float | None = None
